### PR TITLE
Include `extends` keyword in Template

### PR DESCRIPTION
**Pull request by tgodzik**
We already include derive and it makes sense to keep it there.

Fixes https://github.com/scalameta/scalameta/issues/2326

### CODE REVIEW

**Comment by tgodzik** [comment]
@kitbellew It seems the excluding `extend` was intended and it serves a purpose in quasiquotes. I am not sure how much they are currently used and not sure if we can change the behaviour currently. @olafurpg might know more.

We would also need to drop allowing quasiquotes for Mods in constructor params. So quite a few changes for this.

I wonder if we should at some point just extract the parser from scalameta and turn it all legacy. We could drop macros, allow only specific well tested features and cross publish for Scala 3 also. We could also experiment much more without fear of breaking changes. Any change here can have quite a large consequences.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 1]
```diff
@@ -3573,7 +3573,7 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
   }
 
   def ctorModifiers(): Option[Mod] = token match {
-    case Unquote() => Some(unquote[Mod])
+    // case Unquote() => Some(unquote[Mod])
```
Comment (tgodzik) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:3576]: We would need to drop this to make `class A $template` work.

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:3576]: Or actually, we can possibly keep this

File: tests/shared/src/test/scala/scala/meta/tests/quasiquotes/ErrorSuite.scala [hunk 2]
```diff
@@ -720,19 +720,21 @@ class ErrorSuite extends FunSuite {
...
-      |<macro>:4: type mismatch when unquoting;
-      | found   : scala.meta.Type.Select
-      | required: scala.meta.Template
-      |Note: This shape of a quasiquote tells scala.meta that you're unquoting a template.
-      |If you'd like to unquote a parent reference, add {} immediately after the unquote.
-      |For more details, see https://github.com/scalameta/scalameta/issues/223.
-      |      q"class C extends $notReallyAParent"
-      |                        ^
-    """.trim.stripMargin)
+      q"class C $notReallyAParent"
```
Comment (kitbellew) [inline tests/shared/src/test/scala/scala/meta/tests/quasiquotes/ErrorSuite.scala:727]: this type of quasiquote is probably meaningless. as a result of your [welcome] change, there's no more confusion between matching a template and a parent class.

anyone who uses `$foo` without `extends` would know it matches a template. and using `extends $foo` would now require `init"..."` and nothing else.

so i would do this:
- in `ConversionMacros.scala:typeMismatchMessage`, replace the entire template-related block (and `MetaTemplate` field) and remove the `wholeMessage` variable as well.
- change this `#448` test into two:
  - `q"class C $foo` where `foo = t"..."` (this one); wants template, gets type
  - `q"class C extends $foo` where `foo = t"..."` (original, with `extends`); wants init, gets type

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 3]
```diff
@@ -3573,7 +3573,7 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
   }
 
   def ctorModifiers(): Option[Mod] = token match {
-    case Unquote() => Some(unquote[Mod])
+    case Unquote() if ahead(token.is[LeftParen]) => Some(unquote[Mod])
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:3576]: i am not sure i understand this. does this mean unquoting a mod requires `$(mod)`, with parens?

**Review by kitbellew — CHANGES_REQUESTED** [review]
i think this is actually cleaner, no more confusion and need for `{}` (what do these braces do, anyway?).

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 3 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:3576]: It means as far as I can tell `$mod (` so essentially the constructor should be next.

Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:3576]: ah, so this is not for `private class C(...)`, it's for `class C private(...)`? got it.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 4]
```diff
@@ -4753,28 +4753,25 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
...
-            !token.is[LeftParen] && !token.is[LeftBracket] && !token.is[LeftBrace] &&
-            !token.is[KwWith]
-        )) {
-        unquote[Template]
-      } else {
-        template(afterExtend = true, enumCaseAllowed = owner.isEnum)
-      }
+      template(afterExtend = true, enumCaseAllowed = owner.isEnum)
     } else {
       val startPos = in.tokenPos
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:4767]: this is not used, i think

File: tests/shared/src/test/scala/scala/meta/tests/quasiquotes/SuccessSuite.scala [hunk 5]
```diff
@@ -2317,7 +2317,7 @@ class SuccessSuite extends FunSuite {
   test("#448") {
     val parent = init"_root_.scala.AnyVal"
     val template = template"$parent"
-    assert(q"class C extends $template".syntax == "class C extends _root_.scala.AnyVal")
+    assert(q"class C $template".syntax == "class C extends _root_.scala.AnyVal")
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply tests/shared/src/test/scala/scala/meta/tests/quasiquotes/SuccessSuite.scala:2320]: Added! It wasn't tested anywhere else though.

File: tests/shared/src/test/scala/scala/meta/tests/quasiquotes/ErrorSuite.scala (hunk 2 above)
Comment (tgodzik) [reply tests/shared/src/test/scala/scala/meta/tests/quasiquotes/ErrorSuite.scala:727]: Did both, I think we should be fine now.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 4 above)
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Removed

File: tests/shared/src/test/scala/scala/meta/tests/quasiquotes/SuccessSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (kitbellew) [reply tests/shared/src/test/scala/scala/meta/tests/quasiquotes/SuccessSuite.scala:2320]: because it wasn't possible before your change

**Review by kitbellew — APPROVED** [review]

**Pull request merged**